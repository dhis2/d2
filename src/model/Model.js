import { checkType, hasKeys } from '../lib/check';
import { pickOr } from '../lib/utils';
import ModelBase, { DIRTY_PROPERTY_LIST } from './ModelBase';
import createPropertyDefinitionsForAttributes from './helpers/attibutes';

const pickAttributeValues = pickOr('attributeValues', []);

// TODO: Perhaps we can generate model classes dynamically based on the schemas and inherit from this.
/**
 * @extends ModelBase
 *
 * @description
 * A Model represents an object from the DHIS2 Api. A model is created based of a ModelDefinition. The ModelDefinition
 * has the properties that the model should have.
 *
 * @memberof module:model
 */
class Model extends ModelBase {
    /**
     * @constructor
     *
     * @param {ModelDefinition} modelDefinition The model definition that corresponds with the model.
     * This is essential defining what type the model is representing.
     *
     * @description
     * Will create a new model instanced based on the model definition. When creating a new instance the model
     * definition needs to have both the modelValidations and modelProperties.
     *
     * The model properties will depend on the ModelDefinition. A model definition is based on a DHIS2 Schema.
     */
    constructor(modelDefinition) {
        super();
        checkType(modelDefinition, 'object', 'modelDefinition');
        checkType(modelDefinition.modelProperties, 'object', 'modelProperties');

        /**
         * @property {ModelDefinition} modelDefinition Stores reference to the modelDefinition that was used when
         * creating the model. This property is not enumerable or writable and will therefore not show up when looping
         * over the object properties.
         */
        Object.defineProperty(this, 'modelDefinition', { value: modelDefinition });

        /**
         * @property {Boolean} dirty Represents the state of the model. When the model is concidered `dirty`
         * there are pending changes.
         * This property is not enumerable or writable and will therefore not show up when looping
         * over the object properties.
         */
        Object.defineProperty(this, 'dirty', { writable: true, value: false });

        /**
         * @private
         * @property {Object} dataValues Values object used to store the actual model values. Normally access to the
         * Model data will be done through accessor properties that are generated from the modelDefinition.
         *
         * @note {warning} This should not be accessed directly.
         */
        Object.defineProperty(this, 'dataValues', { configurable: true, writable: true, value: {} });

        /**
         * Attach the modelDefinition modelProperties (the properties from the schema) onto the Model.
         *
         * For a data element model the modelProperties would be the following
         * https://play.dhis2.org/demo/api/schemas/dataElement.json?fields=properties
         */
        Object.defineProperties(this, modelDefinition.modelProperties);

        const attributes = {};
        const attributeProperties = modelDefinition.attributeProperties;
        if (hasKeys(attributeProperties)) {
            /**
             * @property {Object} attributes The attributes objects contains references to custom attributes defined
             * on the metadata object.
             *
             * @description
             * These properties are generated based of the attributes that are created for the the specific schema.
             * As these properties are not defined on the schemas they're put on a separate attributes object.
             * When there are no attributes defined for the object type, the attributes property will not be attached
             * to the model.
             *
             * @see https://docs.dhis2.org/2.27/en/user/html/dhis2_user_manual_en_full.html#manage_attribute
             */
            Object.defineProperty(this, 'attributes', { value: attributes });

            const getAttributeValues = () => pickAttributeValues(this);
            const setAttributeValues = attributeValues => (this.attributeValues = attributeValues);
            const setDirty = () => (this.dirty = true);
            const attributeDefinitions = createPropertyDefinitionsForAttributes(
                attributeProperties,
                getAttributeValues,
                setAttributeValues,
                setDirty,
            );

            Object.defineProperties(attributes, attributeDefinitions);
        }

        this[DIRTY_PROPERTY_LIST] = new Set([]);
    }

    /**
     * @static
     *
     * @param {ModelDefinition} modelDefinition ModelDefinition from which the model should be created
     * @returns {Model} Returns an instance of the model.
     *
     * @description The static method is a factory method to create Model objects. It calls `new Model()` with the passed `ModelDefinition`.
     *
     * ```js
     * let myModel = Model.create(modelDefinition);
     * ```
     */
    static create(modelDefinition) {
        return new Model(modelDefinition);
    }
}

export default Model;
