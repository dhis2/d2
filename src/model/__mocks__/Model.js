export default class Model {}

Model.create = jest.fn(modelDefinition => {
    const model = new Model()

    Object.defineProperty(model, 'modelDefinition', {
        value: modelDefinition,
        enumerable: false,
    })

    Object.defineProperty(model, 'dataValues', {
        value: {},
        enumerable: false,
    })

    const propertyDescriptors = Object.keys(
        modelDefinition.modelProperties
    ).reduce((descriptors, propertyName) => {
        descriptors[propertyName] = {
            // eslint-disable-line no-param-reassign
            set(value) {
                this.dataValues[propertyName] = value
            },

            get() {
                return this.dataValues[propertyName]
            },
            enumerable: true,
        }

        return descriptors
    }, {})

    Object.defineProperties(model, propertyDescriptors)

    return model
})
