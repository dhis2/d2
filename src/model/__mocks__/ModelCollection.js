class ModelCollection {
    constructor(modelDefinition, items = []) {
        items.map(item => [item.id, item])
        this.modelDefinition = modelDefinition
    }
}

ModelCollection.create = jest.fn((...args) => new ModelCollection(...args))

module.exports = ModelCollection
