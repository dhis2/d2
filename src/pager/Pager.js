import { isDefined } from '../lib/check'

/**
 * @description
 * Pager object that can be used to navigate pages within a `Modelcollection`
 *
 * @memberof module:model
 */

/**
 * @memberof module:model
 */
// TODO: Move to the model map
class Pager {
    /**
     * @constructor
     *
     * @param {Object} [pager={page: 1, pageCount: 1}] Paging information object.
     * @param {Object} [pagingHandler={list: () => Promise.reject('No handler available')}] Paging handler object. The requirement for this object is that it has a list method.
     *
     * @description
     * Returns a newly created pager object with methods to navigate pages.
     */
    constructor(
        pager = { page: 1, pageCount: 1, query: {} },
        pagingHandler = { list: () => Promise.reject('No handler available') }
    ) {
        /**
         * @property {number} page Current page number
         */
        this.page = pager.page

        /**
         * @property {number} pageCount The total number of pages available
         */
        this.pageCount = pager.pageCount

        /**
         * @property {number} total The total number of items available.
         *
         * @description
         * This represents the total number of items available in the system. Note it is not the number of items
         * on the current page.
         */
        this.total = pager.total

        /**
         * @property {string} nextPage The url to the next page.
         *
         * @description
         * If there is no next page then this will be undefined.
         */
        this.nextPage = pager.nextPage

        /**
         * @property {string} prevPage The url to the previous page
         *
         * @description
         * If there is no previous page then this will be undefined.
         */
        this.prevPage = pager.prevPage

        /**
         * @property {object} query Query parameters
         *
         * @description
         * Query parameters are used for things like filtering and field selection. Used to guarantee that pages are
         * from the same collection.
         */
        this.query = pager.query

        this.pagingHandler = pagingHandler
    }

    /**
     * @returns {Boolean} Result is true when there is a next page, false when there is not.
     *
     * @description
     * Check whether there is a next page.
     */
    hasNextPage() {
        return isDefined(this.nextPage)
    }

    /**
     * Check whether there is a previous page.
     *
     * @returns {Boolean} Result is true when there is a previous page, false when there is not.
     */
    hasPreviousPage() {
        return isDefined(this.prevPage)
    }

    /**
     * @description
     * Loads the next page in the collection if there is one. If no additional pages are available the Promise will reject.
     *
     * @returns {Promise} Promise that resolves with a new `ModelCollection` containing the next page's data. Or rejects with
     * a string when there is no next page for this collection or when the request for the next page failed.
     *
     * @example
     * d2.models.organisationUnit
     *   .list()
     *   .then(collection => {
     *     collection.pager.getNextPage()
     *       .then(secondPageCollection => {
     *         console.log(secondPageCollection.toArray());
     *       });
     *   });
     */
    getNextPage() {
        if (this.hasNextPage()) {
            return this.goToPage(this.page + 1)
        }
        return Promise.reject('There is no next page for this collection')
    }

    /**
     * @description
     * Loads the previous page in the collection if there is one. If no previous pages are available the Promise will reject.
     *
     * @returns {Promise} Promise that resolves with a new `ModelCollection` containing the previous page's data. Or rejects with
     * a string when there is no previous page for this collection or when the request for the previous page failed.
     *
     * @example
     * d2.models.organisationUnit
     *   .list()
     *   .then(collection => {
     *     collection.pager.goToPage(3)
     *       .then(collection => collection.pager.getPreviousPage())
     *       .then(secondPageCollection => {
     *         console.log(secondPageCollection.toArray());
     *       });
     *   });
     */
    getPreviousPage() {
        if (this.hasPreviousPage()) {
            return this.goToPage(this.page - 1)
        }
        return Promise.reject('There is no previous page for this collection')
    }

    /**
     * Loads a specific page for the collection. If the requested page is out of the range of available pages (e.g < 0 or > page count)
     * the Promise will reject with an error.
     *
     * @param {Number} pageNr The number of the page you wish to navigate to.
     * @returns {Promise} Promise that resolves with a new `ModelCollection` containing the data for the requested page.
     *
     * @example
     * d2.models.organisationUnit
     *   .list()
     *   .then(collection => {
     *     collection.pager.goToPage(4)
     *       .then(fourthPageCollection => {
     *         console.log(fourthPageCollection.toArray());
     *       });
     *   });
     */
    // TODO: Throwing the errors here is not really consistent with the rejection of promises for the getNextPage and getPreviousPage
    goToPage(pageNr) {
        if (pageNr < 1) {
            throw new Error('PageNr can not be less than 1')
        }
        if (pageNr > this.pageCount) {
            throw new Error(
                `PageNr can not be larger than the total page count of ${
                    this.pageCount
                }`
            )
        }

        return this.pagingHandler.list(
            Object.assign({}, this.query, { page: pageNr })
        )
    }
}

export default Pager
