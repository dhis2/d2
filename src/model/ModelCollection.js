import { isValidUid, isArray, checkType } from '../lib/check'
import { throwError } from '../lib/utils'
import Model from './Model'
import ModelDefinition from './ModelDefinition'
import Pager from '../pager/Pager'

function throwIfContainsOtherThanModelObjects(values) {
    if (values && values[Symbol.iterator]) {
        const toCheck = [...values]
        toCheck.forEach(value => {
            if (!(value instanceof Model)) {
                throwError(
                    'Values of a ModelCollection must be instances of Model'
                )
            }
        })
    }
}

function throwIfContainsModelWithoutUid(values) {
    if (values && values[Symbol.iterator]) {
        const toCheck = [...values]
        toCheck.forEach(value => {
            if (!isValidUid(value.id)) {
                throwError(
                    'Can not add a Model without id to a ModelCollection'
                )
            }
        })
    }
}

/**
 * Collection of `Model` objects that can be interacted upon. Can contain a pager object to easily navigate
 * pages within the system.
 *
 * @memberof module:model
 */
class ModelCollection {
    /**
     * @constructor
     *
     * @param {ModelDefinition} modelDefinition The `ModelDefinition` that this collection is for. This defines the type of models that
     * are allowed to be added to the collection.
     * @param {Model[]} values Initial values that should be added to the collection.
     * @param {Object} pagerData Object with pager data. This object contains data that will be put into the `Pager` instance.
     *
     * @description
     *
     * Creates a new `ModelCollection` object based on the passed `modelDefinition`. Additionally values can be added by passing
     * `Model` objects in the `values` parameter. The collection also exposes a pager object which can be used to navigate through
     * the pages in the collection. For more information see the `Pager` class.
     */
    constructor(modelDefinition, values, pagerData) {
        checkType(modelDefinition, ModelDefinition)
        /**
         * @property {ModelDefinition} modelDefinition The `ModelDefinition` that this collection is for. This defines the type of models that
         * are allowed to be added to the collection.
         */
        this.modelDefinition = modelDefinition

        /**
         * @property {Pager} pager Pager object that is created from the pagerData that was passed when the collection was constructed. If no pager data was present
         * the pager will have default values.
         */
        this.pager = new Pager(pagerData, modelDefinition)

        // We can not extend the Map object right away in v8 contexts.
        this.valuesContainerMap = new Map()
        this[Symbol.iterator] = this.valuesContainerMap[Symbol.iterator].bind(
            this.valuesContainerMap
        )

        throwIfContainsOtherThanModelObjects(values)
        throwIfContainsModelWithoutUid(values)

        // Add the values separately as not all Iterators return the same values
        if (isArray(values)) {
            values.forEach(value =>
                this.valuesContainerMap.set(value.id, value)
            )
        }
    }

    /**
     * @property {Number} size The number of Model objects that are in the collection.
     *
     * @description
     * Contains the number of Model objects that are in this collection. If the collection is a collection with a pager. This
     * does not take into account all the items in the database. Therefore when a pager is present on the collection
     * the size will return the items on that page. To get the total number of items consult the pager.
     */
    get size() {
        return this.valuesContainerMap.size
    }

    /**
     * Adds a Model instance to the collection. The model is checked if it is a correct instance of `Model` and if it has
     * a valid id. A valid id is a uid string of 11 alphanumeric characters.
     *
     * @param {Model} value Model instance to add to the collection.
     * @returns {ModelCollection} Returns itself for chaining purposes.
     *
     * @throws {Error} When the passed value is not an instance of `Model`
     * @throws {Error} Throws error when the passed value does not have a valid id.
     */
    add(value) {
        throwIfContainsOtherThanModelObjects([value])
        throwIfContainsModelWithoutUid([value])

        this.set(value.id, value)
        return this
    }

    /**
     * If working with the Map type object is inconvenient this method can be used to return the values
     * of the collection as an Array object.
     *
     * @returns {Array} Returns the values of the collection as an array.
     */
    toArray() {
        const resultArray = []

        this.forEach(model => {
            resultArray.push(model)
        })

        return resultArray
    }

    static create(modelDefinition, values, pagerData) {
        return new ModelCollection(modelDefinition, values, pagerData)
    }

    static throwIfContainsOtherThanModelObjects(value) {
        return throwIfContainsOtherThanModelObjects(value)
    }

    static throwIfContainsModelWithoutUid(value) {
        return throwIfContainsModelWithoutUid(value)
    }

    /**
     * Clear the collection and remove all it's values.
     *
     * @returns {this} Returns itself for chaining purposes;
     */
    // TODO: Reset the pager?
    clear() {
        return this.valuesContainerMap.clear.call(this.valuesContainerMap)
    }

    delete(...args) {
        return this.valuesContainerMap.delete.call(
            this.valuesContainerMap,
            ...args
        )
    }

    entries() {
        return this.valuesContainerMap.entries.call(this.valuesContainerMap)
    }

    // FIXME: This calls the forEach function with the values Map and not with the ModelCollection as the third argument
    forEach(...args) {
        return this.valuesContainerMap.forEach.call(
            this.valuesContainerMap,
            ...args
        )
    }

    get(...args) {
        return this.valuesContainerMap.get.call(
            this.valuesContainerMap,
            ...args
        )
    }

    has(...args) {
        return this.valuesContainerMap.has.call(
            this.valuesContainerMap,
            ...args
        )
    }

    keys() {
        return this.valuesContainerMap.keys.call(this.valuesContainerMap)
    }

    set(...args) {
        return this.valuesContainerMap.set.call(
            this.valuesContainerMap,
            ...args
        )
    }

    values() {
        return this.valuesContainerMap.values.call(this.valuesContainerMap)
    }
}

export default ModelCollection
