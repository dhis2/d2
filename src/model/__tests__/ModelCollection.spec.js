import Model from '../Model'
import ModelDefinition from '../ModelDefinition'
import Pager from '../../pager/Pager'
import ModelCollection from '../ModelCollection'

jest.mock('../Model')
jest.mock('../../../src/pager/Pager')

describe('ModelCollection', () => {
    const mockSchema = { singular: 'mock', plural: 'mocks' }
    const mockModelDefinition = new ModelDefinition(mockSchema, [])
    let pagerObject

    beforeEach(() => {
        pagerObject = {
            page: 1,
            pageCount: 10,
            total: 482,
            nextPage: 'http://localhost:8080/dhis/api/dataElements?page=2',
        }

        Pager.mockReset()
        Pager.mockReturnValueOnce(pagerObject)
    })

    describe('extension of Map', () => {
        let firstValue
        let modelCollection

        beforeEach(() => {
            firstValue = new Model(mockModelDefinition)
            firstValue.id = 'q2egwkkrfco'

            modelCollection = ModelCollection.create(
                mockModelDefinition,
                [],
                pagerObject
            )
            modelCollection.add(firstValue)
        })

        it('should have a clear method that clears the list', () => {
            modelCollection.clear()

            expect(modelCollection.size).toBe(0)
        })

        it('should get the values', () => {
            expect(Array.from(modelCollection.values())[0]).toBe(firstValue)
        })

        it('should get the keys', () => {
            expect(Array.from(modelCollection.keys())[0]).toBe('q2egwkkrfco')
        })

        it('should get the entries', () => {
            expect(modelCollection.entries().next().value).toEqual([
                firstValue.id,
                firstValue,
            ])
        })

        it('should run the forEach with the correct values', () => {
            const forEachFunc = jest.fn()

            modelCollection.forEach(forEachFunc)

            expect(forEachFunc).toBeCalledWith(
                firstValue,
                'q2egwkkrfco',
                modelCollection.valuesContainerMap
            )
        })

        it('should remove the correct value', () => {
            modelCollection.delete('q2egwkkrfco')

            expect(modelCollection.size).toBe(0)
        })

        it('should get the entries', () => {
            expect(Array.from(modelCollection)[0]).toEqual([
                'q2egwkkrfco',
                firstValue,
            ])
        })

        it('should return true when the entry is in the collection', () => {
            expect(modelCollection.has('q2egwkkrfco')).toBe(true)
        })

        it('should return the correct value on get', () => {
            expect(modelCollection.get('q2egwkkrfco')).toBe(firstValue)
        })

        it('should throw error when trying to set the size', () => {
            expect(() => {
                modelCollection.size = 0
            }).toThrowError()
        })
    })

    it('should be an object', () => {
        expect(ModelCollection).toBeInstanceOf(Function)
    })

    it('should accept 3 arguments', () => {
        expect(ModelCollection.length).toBe(3)
    })

    describe('class', () => {
        describe('create method', () => {
            it('should be a function', () => {
                expect(ModelCollection.create).toBeInstanceOf(Function)
            })

            it('should return an instance of the class', () => {
                expect(
                    ModelCollection.create(mockModelDefinition)
                ).toBeInstanceOf(ModelCollection)
            })

            it('should instantiate a new pager', () => {
                const collection = ModelCollection.create(mockModelDefinition)

                expect(collection.pager).toBeDefined()
            })

            it('should not be allowed to be called without new', () => {
                expect(() => ModelCollection()).toThrowErrorMatchingSnapshot()
            })
        })

        describe('throwIfContainsOtherThanModelObjects', () => {
            it('should throw when one of the the passed values in the array is not a Model', () => {
                expect(() =>
                    ModelCollection.throwIfContainsOtherThanModelObjects([{}])
                ).toThrowError(
                    'Values of a ModelCollection must be instances of Model'
                )
            })

            it('should not throw when the passed value is a model', () => {
                expect(() =>
                    ModelCollection.throwIfContainsOtherThanModelObjects([
                        new Model(mockModelDefinition),
                    ])
                ).not.toThrowError()
            })
        })

        describe('throwIfContainsModelWithoutUid', () => {
            it('should throw when the passed array contains a modelWithoutId', () => {
                expect(() =>
                    ModelCollection.throwIfContainsModelWithoutUid([
                        new Model(mockModelDefinition),
                    ])
                ).toThrowError(
                    'Can not add a Model without id to a ModelCollection'
                )
            })

            it('should accept models with valid UIDs', () => {
                const model = new Model(mockModelDefinition)

                model.id = 'FQ2o8UBlcrS'

                expect(() =>
                    ModelCollection.throwIfContainsModelWithoutUid([model])
                ).not.toThrowError()
            })
        })
    })

    describe('instance', () => {
        let modelDefinition
        let modelCollection
        let mockyModel1
        let mockyModel2
        let mockyModel3

        beforeEach(() => {
            modelDefinition = new ModelDefinition(mockSchema, [])
            modelCollection = new ModelCollection(modelDefinition)
            mockyModel1 = new Model(mockModelDefinition)
            mockyModel1.id = 'q2egwkkrfc1'
            mockyModel2 = new Model(mockModelDefinition)
            mockyModel2.id = 'q2egwkkrfc2'
            mockyModel3 = new Model(mockModelDefinition)
            mockyModel3.id = 'q2egwkkrfc3'
        })

        it('should throw if being constructed with non Model values', () => {
            expect(
                () => new ModelCollection(modelDefinition, [1, 2, 3])
            ).toThrowError(
                'Values of a ModelCollection must be instances of Model'
            )
        })

        it('should accept an array of Model objects', () => {
            modelCollection = new ModelCollection(modelDefinition, [
                mockyModel1,
                mockyModel2,
                mockyModel3,
            ])

            expect(modelCollection.size).toBe(3)
        })

        it('should not add the same model twice', () => {
            modelCollection = new ModelCollection(modelDefinition, [
                mockyModel1,
                mockyModel1,
            ])

            expect(modelCollection.size).toBe(1)
        })

        it('should return the first Model', () => {
            const firstModel = mockyModel1
            mockyModel2.id = firstModel.id

            modelCollection = new ModelCollection(modelDefinition, [
                firstModel,
                mockyModel2,
            ])

            const firstValue = modelCollection.get('q2egwkkrfc1')

            expect(modelCollection.size).toBe(1)
            expect(firstValue.id).toBe('q2egwkkrfc1')
            expect(firstValue).toBeInstanceOf(Model)
        })

        it('should set the modelDefinition onto the modelCollection', () => {
            expect(modelCollection.modelDefinition).toBe(modelDefinition)
        })

        describe('add', () => {
            it('should accept Model as a value', () => {
                modelCollection.add(mockyModel1)
            })

            it('should not accept a number', () => {
                expect(() => modelCollection.add(1)).toThrowError(
                    'Values of a ModelCollection must be instances of Model'
                )
            })

            it('should not accept an empty object', () => {
                expect(() => modelCollection.add({})).toThrowError(
                    'Values of a ModelCollection must be instances of Model'
                )
            })

            it('should not accept an object that was created based on a local class', () => {
                class Model {
                    constructor(id) {
                        this.id = id
                    }
                }

                expect(() =>
                    modelCollection.add(new Model('q2egwkkrfco'))
                ).toThrowError(
                    'Values of a ModelCollection must be instances of Model'
                )
            })

            it('should accept an object that was create with Model as subclass', () => {
                class MyModel extends Model {}
                const myModel = new MyModel(mockModelDefinition)
                myModel.id = 'q2egwkkrfco'

                expect(() => modelCollection.add(myModel)).not.toThrowError(
                    'Values of a ModelCollection must be instances of Model'
                )
                expect(modelCollection.size).toBe(1)
            })

            it('should throw if the id is not available', () => {
                expect(() =>
                    modelCollection.add(new Model(mockModelDefinition))
                ).toThrowError(
                    'Can not add a Model without id to a ModelCollection'
                )
            })
        })

        describe('toArray', () => {
            it('should return an array of the items', () => {
                const modelArray = [mockyModel1, mockyModel2]

                modelCollection = new ModelCollection(
                    modelDefinition,
                    modelArray
                )

                expect(modelCollection.toArray()).toEqual(modelArray)
            })
        })
    })
})
