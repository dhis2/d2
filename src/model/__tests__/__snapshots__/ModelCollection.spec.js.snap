// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelCollection class create method should not be allowed to be called without new 1`] = `"Class constructor ModelCollection cannot be invoked without 'new'"`;
